overlord/servicestate/quota_handlers.go

```diff
@@ -37,6 +40,302 @@ import (
 	"github.com/snapcore/snapd/wrappers"
 )
 
+// QuotaControlAction is the serialized representation of a quota group
+// modification that lives in a task.
+type QuotaControlAction struct {
+	// QuotaName is the name of the quota group being controlled.
+	QuotaName string `json:"quota-name"`
+
+	// Action is the action being taken on the quota group. It can be either
+	// "create", "update", or "remove".
+	Action string `json:"action"`
+
+	// AddSnaps is the set of snaps to add to the quota group, valid for either
+	// the "update" or the "create" actions.
+	AddSnaps []string `json:"snaps"`
+
+	// MemoryLimit is the memory limit for the quota group being controlled,
+	// either the initial limit the group is created with for the "create"
+	// action, or if non-zero for the "update" the memory limit, then the new
+	// value to be set.
+	MemoryLimit quantity.Size
+
+	// ParentName is the name of the parent for the quota group if it is being
+	// created. Eventually this could be used with the "update" action to
+	// support moving quota groups from one parent to another, but that is
+	// currently not supported.
+	ParentName string
+}
+
+func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	meter := snapstate.NewTaskProgressAdapterUnlocked(t)
+
+	qcs := []QuotaControlAction{}
+	err := t.Get("quota-control-actions", &qcs)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get quota-control-action: %v", err)
+	}
+
+	// TODO: support more than one action
+	switch {
+	case len(qcs) > 1:
+		return fmt.Errorf("multiple quota group actions not supported yet")
+	case len(qcs) == 0:
+		return fmt.Errorf("internal error: no quota group actions for quota-control task")
+	}
+
+	qc := qcs[0]
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	args := quotaControlArgs{
+		St:          st,
+		AllGrps:     allGrps,
+		Action:      &qc,
+		Meter:       meter,
+		PerfTimings: perfTimings,
+	}
+
+	switch qc.Action {
+	case "create":
+		err = doCreateQuota(args)
+	case "remove":
+		err = doRemoveQuota(args)
+	case "update":
+		err = doUpdateQuota(args)
+	default:
+		err = fmt.Errorf("unknown action %q requested", qc.Action)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	t.SetStatus(state.DoneStatus)
+
+	return nil
+}
+
+func doCreateQuota(args quotaControlArgs) error {
+	st := args.St
```

Comment:
I'm not a fan of using structs to encompass all the args, in particular *state.State, it's useful to see at a glance if a function needs state so I'd leave it as an explicit 1st argument. Same for meter and timings (we typically have them last as they are least important).

Reply:
seems my attempts at making function signatures shorter are now 0/2 oh well

Reply:
removed the args struct